This is the first of a series of tutorials on useful R packges which I call the lifesaver R packages. If you want to run the tutorials jupyter, you can find the notebooks at [lifesaver-R-packages](https://github.com/shettysindhu63/lifesaver-R-packages). 

Today we will be exploring the titanic dataset with two of my favourite libraries, **ggplot2** and **dplyr**. For more details about the dataset, visit the [Kaggle site](https://www.kaggle.com/c/titanic).

### The Packages 

- The **dplyr** package provides functions for easy and fast data manipulation. You can review the [dplyr vignette](https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html) for a comprehensive introduction to **dplyr** verbs. 


- **ggplot2** is a graphing package implemented in R based on the [Grammar of Graphics](http://www.springer.com/gp/book/9780387245447). In the grammar of graphics system, the graphic/plot is made up of several layers : data, aesthetics, geometries, facets, statistics, coordinates and theme. A compact intro can be read [here](https://codewords.recurse.com/issues/six/telling-stories-with-data-using-the-grammar-of-graphics).

### Get the data

The first step is to import the required packages and the datasets. 
Along with **ggplot2** and **dplyr**, we will also use **magrittr** which lets us use the pipe operation that makes the code more readable and concise. 

In [47]:
# load packages 
library(ggplot2)
library(dplyr)
library(magrittr)

In [48]:
# load data
dataset = read.csv('data/titanic/train.csv')

# check for missing values 
dataset %>% 
    sapply(function(y) sum(is.na(y)))

PassengerId    Survived      Pclass        Name         Sex         Age 
          0           0           0           0           0         177 
      SibSp       Parch      Ticket        Fare       Cabin    Embarked 
          0           0           0           0           0           0

We see that the value for column Age is missing in 177 observations. There are several ways to handle missing values: you could omit them completely or impute the missing values with mean, median or some other statistic of all other values in that partcular column. In this case, I do not want to presume anything about the ages of the passengers, so I will exclude the missing observations completely from further analysis which includes the Age feature.

Let's look at the first few rows of the dataset while we decide what we might want to explore. 

In [49]:
# print first six rows of the dataset
head(dataset)

PassengerId Survived Pclass
1 1           0        3     
2 2           1        1     
3 3           1        3     
4 4           1        1     
5 5           0        3     
6 6           0        3     
  Name                                                Sex    Age SibSp Parch
1 Braund, Mr. Owen Harris                             male   22  1     0    
2 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38  1     0    
3 Heikkinen, Miss. Laina                              female 26  0     0    
4 Futrelle, Mrs. Jacques Heath (Lily May Peel)        female 35  1     0    
5 Allen, Mr. William Henry                            male   35  0     0    
6 Moran, Mr. James                                    male   NA  0     0    
  Ticket           Fare    Cabin Embarked
1 A/5 21171         7.2500       S       
2 PC 17599         71.2833 C85   C       
3 STON/O2. 3101282  7.9250       S       
4 113803           53.1000 C123  S       
5 373450            8.0500       S       
6 330877            8.4583       Q

I suggest looking into the [data overview](https://www.kaggle.com/c/titanic/data) provided by Kaggle to check what each of the columns represent. Some columns are obviously not relevant to whether the passenger survived or not, like PassengerID, Name, Ticket number. Out of the others, let us select a few columns first to start simple : consider only ticket class (Pclass), sex, age and port of embarkation (Embarked).  

In [50]:
# select only required columns 
dataset = select(dataset, Survived, Pclass, Sex, Age, Embarked)

### Summarize the data 

Next, we may want to see the survival rate for different groups. This is where **dplyr** and **magrittr** together shine, you'll see how easy it is to write the *split-apply-combine* strategy. 

In [63]:
# Check the survival rate for different genders 
dataset %>%
    group_by(Sex) %>%
    summarize(Total = length(Sex), Survived  = sum(Survived), Rate = round((Survived/Total)*100,2))

Sex    Total Survived Rate 
1 female 314   233      74.20
2 male   577   109      18.89

In [65]:
# Check the survival rate for different classes 
dataset %>%
    group_by(Pclass) %>%
    summarize(Total = length(Pclass), Survived  = sum(Survived), Rate = round((Survived/Total)*100,2))

Pclass Total Survived Rate 
1 1      216   136      62.96
2 2      184    87      47.28
3 3      491   119      24.24

In [67]:
# Check the survival rate by grouping Sex and Pclass
dataset %>%
    group_by(Sex, Pclass) %>%
    summarize(Total = length(Pclass), Survived  = sum(Survived), Rate = round((Survived/Total)*100,2))

Sex    Pclass Total Survived Rate 
1 female 1       94   91       96.81
2 female 2       76   70       92.11
3 female 3      144   72       50.00
4 male   1      122   45       36.89
5 male   2      108   17       15.74
6 male   3      347   47       13.54

###  Visualize the data 

As the tables get bigger, it will become difficult to interpret the numbers. This is where exploratory graphs can help us understand the data better. 

In [72]:
# visualising the survival rate for sex and Pclass groups


### References 

1. [dplyr CRAN vignette](https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html)
2. [magrittr CRAN vignette](https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html)
3. [tidyverse ggplot2 reference](http://ggplot2.tidyverse.org/reference/)
4. [Telling stories with data using the grammar of graphics](https://codewords.recurse.com/issues/six/telling-stories-with-data-using-the-grammar-of-graphics), Liz Sander, Code Words - Issue Six